### ISSUE TITLE
Allow authors to be empty?

### ISSUE THREAD

**Issue by ioquatix** [rubygems/rubygems#5358]
https://github.com/rubygems/rubygems/blob/4b0d970cd31a7a34016bb6668f9756466070d6aa/lib/rubygems/specification_policy.rb#L346-L350

In some cases, it may not be possible to list all authors. So a subset is given. We would like to encourage people in this case to refer to an external file (e.g. LICENSE) to list all authors/contributors. Whether this is in scope for the gem spec, I don't know. However, because of this requirement, something must be listed.

I'd like to know, can we drop this requirement?

**Comment by ioquatix** [comment]
To give a specific example, if `rails` listed all authors, (4000+) it could add several 10s of KiB to the gem specification just in strings alone.

**Comment by simi** [comment]
I'm not sure this is really issue. Anyway "authors" field is not "contributors" field in my mind. But I don't see any problem write a note to the authors field like "Rails team".

**Comment by ioquatix** [comment]
What is the difference between an author and a contributor?

**Comment by simi** [comment]
> What is the difference between an author and a contributor?

I'm not going to explain on this since it is highly subjective.

---

Anyway feel free to check if removing author presence would cause any troubles. I can imagine there is some code relying on presence of author and removing the validation could broke it (even outside of RubyGems). As I mentioned before, you're free to put anything generic in there like following.

```ruby
spec.author = 'Rack maintainers'
```

We can improve https://guides.rubygems.org/specification-reference/#authors= as well if you think that's not descriptive enough.

**Comment by deivid-rodriguez** [comment]
I'm also now sure this is a problem.

Is someone listing too many authors in this field, so that we need to take any actions to prevent it?

**Comment by ioquatix** [comment]
Well, since you asked, Rails has 4000+ authors/contributors, so it's too much for the field IMHO :)

I'd personally prefer to remove author from my gem (i.e. empty field). Otherwise I can just add a place holder e.g. `spec.author = "No-one"` or `spec.author = "My Team"` but my preference is to just avoid it entirely.

An empty array should be okay, IMHO most use cases would be something like `spec.authors.each ...` so it won't hurt if it's empty.

**Comment by simi** [comment]
> Well, since you asked, Rails has 4000+ authors/contributors, so it's too much for the field IMHO :) I can try publishing a gem with 4000 authors and we can see what happens? What about 40,000 authors? Is there an upper limit? Do we do anything with that field in RG.org?

@ioquatix But that's totally made up scenario. Rails doesn't do that and RubyGems doesn't suggest do anything like that (list everyone contributing to the project in the gemspec). In general the same problem is present for all array values like `files`. We can provide warning when there is more than x values in `validate_array_attribute` in `SpecificationPolicy` to prevent this. We can also in general provide warning if gemspec size reaches some threshold.

**Comment by deivid-rodriguez** [comment]
Exactly, I was asking for a real scenario, not scenario, not something made up. Obviously someone could actually do this, but it hasn't happened in like 15 years, or if it has actually happened, nobody has reported any issues associated with doing it. To me, this feels like fixing a non issue. If we want to improve this, any of the more generic improvements suggested by @simi feels better.

If your rationale for making this field optional is something else, then that'd be a separate discussion.

**Comment by ioquatix** [comment]
I'm less interested in the 4000 authors case, but if you check Rails, they do actually have 4000+ contributors and in theory they are all authors so that wouldn't be unreasonable to have a list of authors 4000 long, it would be easy to automate that via `git log` etc.

The alternative is you choose one of the following:

1. Use the first person who created the gem as the author and ignore everyone else.
2. List all people who contributed to a gem and list them as authors (feels impractical after more than, say, 10 authors).
3. Use a place holder like "Team X" or some other made up name.
4. Don't list any authors.

It's option 4 I'd like to support since option 1 seems unfair to other authors/contributors, option 2 won't work generally in the case there are even 10s or 100s of authors/contributors and option 3 feels like an abuse of the field.

**Comment by ioquatix** [comment]
I'd be interested to take a look at gems with long lists of authors, I believe bundler appears to follow (2), e.g. https://rubygems.org/gems/bundler has a long list of authors but I'm not sure if it's complete, I assume not. Who decides who is an author or not?

**Comment by simi** [comment]
@ioquatix I think you're looking for problems where actually no problems are present. There are no rules how authors gemspec metadata attribute works. You put something in there and it will appear at https://rubygems.org/gems/rails at authors section and also it is visible at gem info output. That's it.

> Who decides who is an author or not?

The one filling in gemspec data.

**Comment by ioquatix** [comment]
Right, but in my gem specs I don't want to list any author.

**Comment by simi** [comment]
Than I suggest to build gem using `--force` option and distribute it outside of the rubygems.org.

**Comment by ioquatix** [comment]
What are the technical reason why `authors` cannot be an empty array?

**Comment by simi** [comment]
There are mostly no technical reasons. You can probably bypass the validation using `--force`. There are some validations on rubygems.org side, feel free to check if gem with missing authors will be accepted in there. Anyway I don't see any real reason to make author empty and you have not provided any real use-case. We can in theory make almost whole gemspec optional (except name and files). But what would be the benefit? Majority of gems are hosted at rubygems.org which is public catalog of gems and it is nice to have various metadata around to enrich this catalog.

To summarise - I don't think this is good idea and I'm not going to participate here anymore. I suggest to close this and https://github.com/rubygems/rubygems/pull/5362. If you would like to push it forward anyway, good next step would be to check if rubygems.org would accept this kind of gem and contribute in there if there will be any problems found regarding this change.

**Comment by ioquatix** [comment]
I'm experimenting with setting the authors field to be the list of all authors from `git log`.

https://github.com/ioquatix/bake-modernize/blob/63bd3811b8c64b44dafefbb482d69bec2f744a31/bake/modernize/gemspec.rb#L194-L211

I'm okay with this approach and I think it's the most logical given that if we want to present authors, it should be a complete list.

I tested out RubyGems.org and found there is a hard limit of 64_000 characters on a single author, or if an array of authors given, the syntax is also included, so it appears the limit is 64_000 characters of `spec.authors.inspect`. Note that this isn't the bytesize.

Not sure how to describe my use case any more clearly. I'm essentially saying:

- Either we list all authors (maybe not acceptable for a variety of reasons), or
- We don't list any authors and users should check git logs, license, or other source of truth.

The motivation is to avoid having a partial list which feels unfair to contributors who are not listed or having a place-holder string like "Team X" (I actually tested `spec.author = ""` which is accepted oddly enough) which seems superfluous.

Within he above logic, listing all authors of some projects will be totally impractical, so the only alternative is either listing none of them, or having a place holder like "Team X". This PR is specifically about this case, and trying to avoid writing `spec.author = "Team X"` in the case that there are literally thousands of authors (which is a valid case as shown by Rails) and possibly other projects, e.g. Rack which has 100s of contributors at least).

**Comment by ioquatix** [comment]
I checked https://github.com/rubygems/rubygems.org/blob/1b269fbb37f45a9769952d0f17b6656ff53e5473/app/views/rubygems/_gem_members.html.erb#L44-L51 and it already looks like it should be okay, since it checks whether authors is present before displaying that entire section.
